### ISSUE TITLE
Word before a quotation mark will not be hyphenated

### ISSUE THREAD

**Issue by grabi678** [koreader/koreader#5645]
* KOReader version: v2019.11
* Device: PocketBook Inkpad 3 (PB740)

#### Issue
The first word in a line before a quotation mark is not hyphenated even if there is more than enough space (German hyphenation).
![IMG_20191126_15112](https://user-images.githubusercontent.com/58222328/69642499-bf8c5400-1059-11ea-8c8d-50cedf51c221.png)
![IMG_20191126_150001915](https://user-images.githubusercontent.com/58222328/69642527-c6b36200-1059-11ea-8145-0be2c5cc9286.png)

**Comment by poire-z** [comment]
> even if there is more than enough space 

(And there's more than enough space to put `diesem` on previous line I guess :)

Not pocketbook specific.
But related to https://github.com/koreader/crengine/issues/307#issuecomment-523920580 - and the fact that we hardcoded the french rules for guillemets. Without that, your` »Warum«` might have been at start of line - but the french rules make the whole combo considered a single word I guess.
Another problem (for me to look at) is why the previous word has not been hyphenated. Might be a bug with that _prevent a line wrap on space surrounded by guillemets and others_ code.

Can't quickly be fixed as we'd need some language detection/selection features to trigger different rules. That's planned in that https://github.com/koreader/crengine/issues/307, but it won't come soon.
Current solution would be to compile a libcrengine.so with these rules reverted. If you can/want to do that, the lines at top and bottom in the following should have the flags inverted:
https://github.com/koreader/crengine/blob/cc9af0ad913dc1357bd968dd503096cbb1fbbdbc/crengine/src/lvstring.cpp#L3501-L3518

**Comment by NiLuJe** [comment]
My German is extremely rusty, but I seem to recall it having some fairly peculiar hyphenation rules, despite its love of gigantically long words...

**Comment by Frenzie** [comment]
Peculiar?

die|sem
ab|zu|schrub|ben

**Comment by poire-z** [comment]
Mhhh, strange, because I can get the word before or after such guillemets hyphenated, with both French and German hyphenation. So, that's somehow reassuring.

With german hyphenation and the text copied from the screenshot, when increasing font size, there are sizes where the hyphenation happens - but in doesn't for some intermediate value !

<kbd>![image](https://user-images.githubusercontent.com/24273478/69649684-b311f680-106d-11ea-985e-394e5e091876.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/69649773-d9379680-106d-11ea-8d75-0671c8b7f716.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/69649820-eb193980-106d-11ea-87b9-0cba0ee1d6b4.png)</kbd>

If the 2nd could have `diesem` hyphenated, the 1st one should have it too.

**Comment by Frenzie** [comment]
Makes you wonder if there somehow might be a non-breaking space before the `»`. Could you share an HTML fragment, @grabi678?

**Comment by poire-z** [comment]
There was plain regular space in my sample.
And the non-breaking space feeling is simulated by my french guillemets rules.

**Comment by poire-z** [comment]
> Could you share an HTML fragment, 

Yes, just to be sure. No need to dig into the EPUB. Just highlight a few wordfs before up to a few words after, then View HTML > Switch to debug view > Switch to extended debug view, and take a photo like you did showing whats around `diesem Warum`.

**Comment by poire-z** [comment]
I guess it's some edge case.
<kbd>![image](https://user-images.githubusercontent.com/24273478/69653875-2585d500-1074-11ea-8500-148be3766e0a.png)</kbd>
In this, there was enough room to put `diesem` on the first line, so we pass by it.
Then, the avoid_wrap_after/before rules decide a wrap should be prevented.
The code that looks for stuff to hyphenate (wrongly, in this case) looks to hyphenate what's after what we pass by (Warum), and no part of it will fit. So, the last ok wrap was before `diesem`.
If diesem was just a bit too large to fit, it would have been considered for hyphenation.

**Comment by Frenzie** [comment]
And that's because of the non-breaking space with guillemets?

**Comment by poire-z** [comment]
Looks like the guillemets and the _avoid wrap on space depending on what's on the left or right_ code is not even the problem.

With no such guillemets, and just a `&nbsp;` between `diesem` and `Warum`:

If the code stops the normal layout (because max line width reached) in the middle of `Warum`, it will try to hyphenate `Warum`. If only `W` would fit, and `Wa` would not, or if the word (`Wzrzm` below) can't be hyphenated, it currently give up and just use lastNormalWrap (before `diesem`).
I think the logic is that it should try to continue hyphenating all the words back until lastNormalWrap: it would then try `diesem`, and would find a hyphWrap at `die-sem` that would fit, and would be used.

<kbd>![image](https://user-images.githubusercontent.com/24273478/69661830-65a08400-1083-11ea-9945-d18965265eff.png)</kbd>

**Comment by grabi678** [comment]
As requested here comes the original code from xhtml files taken from Calibre e-book editing engine. Nothing special in there.
![diesem 'Warum'](https://user-images.githubusercontent.com/58222328/69663480-3cc9c000-107e-11ea-9abd-ea7476a86e96.png)
![abzuschrubben  'Kannst'](https://user-images.githubusercontent.com/58222328/69663488-3f2c1a00-107e-11ea-9584-a553941fe073.png)

**Comment by poire-z** [comment]
Should be fixed with https://github.com/koreader/crengine/pull/320.

@grabi678 : you didn't mention the fact that you don't see a valid line wrap between `diesem` and `»Warum`. You must have never seen a `»` at start of line with KOReader, for the reason explained above. Have you noticed it? Or has that never bothered you?
On a pain scale of 0 to 10, how would you rate this problem with `»` ? (Let's say total garbage would be 10, none or wrong hyphenation would be 9, and this issue with missed hyphenation opportunities would be 7 :)

@ptrm @robert00s same question with Polish and the issue mentionned by @robert00s at https://github.com/koreader/crengine/issues/307#issuecomment-523896177.
On a pain scale of 0 to 10, with the above standard, how would you rate these issues with Polish:
- A) spaces are too wide (just to get another number for comparing pain levels :)
- B) _Polish and Czech typography rules requires avoiding one letter prepositions at line endings_ not ensured (when the publisher has not put the right `&nbsp;` to help us with that)
- C) hyphens not duplicated at start of line (which I read somewhere it must happen in Polish in some cases).

Are these 2 polish rules "must have", or are they really just for perfect typography, and newspapers would not even really care?
Also, for these one letter at end of line, are there a fixed set of such letters, and it would always work with them? Or would you need to know the grammar to distinguish if it's a preposition, or whether it's not a preposition (in which case it is allowed at end of line)?

**Comment by ptrm** [comment]
A: after the floor › round fix it's **9** (painful not garbage), before (current stable version I think), **7**. This scale's curve might be very exponential ;)

B: Haven't read a book without `&nbsp;` placed where needed in a long time (this goes for English (Picador, Vintage, Penguin publishers recently), too) Besides, from what I know, it's not a strict rule but rather a good-to-have, although painful if absent. So **3**. (But I guess since all modern typesetting apps automatically wrap single letters to a new line, it's not commonly seen)

C: example please :D ?

Note for #320: I suspected not all words would break where they would fit partially in the previous line, though never bothered to measure that, thanks :grinning: Plus I wonder how #320 will affect my perception of #319 changes ;D - was not sure where to put this comment ;>
Also, #320 will affect A, I presume.

**Comment by poire-z** [comment]
320 should help with really rare cases (explicite non breaking space, compound words like "aujourd'hui" that could be missed in some cases), but in most cases it shouldn't really have any impact (if I didn't mess it up).
Here's [libcrengine+cre-spaceWidthScale+320forKobo.zip](https://github.com/koreader/koreader/files/3898088/libcrengine%2Bcre-spaceWidthScale%2B320.zip) if you want to confirm it does not change much.

> C: example please :D ?

Ok, may be not generated hyphen from hyphenating words, but explicit hyphens, some example at https://unicode.org/reports/tr14/#Hyphen

> B: Haven't read a book without `&nbsp;`

OK, good. Can you tell me just a bit more if all/subset of letters, and preposition only/always ? Just so that I get these questions cleared for future thinking :)

**Comment by ptrm** [comment]
B: https://sjp.pwn.pl/poradnia/haslo/Sieroty-na-koncu-wiersza;18684.html - that's, I think, our official body of language prescriptionists that responds here.
Translation screenshot (Google managed to maintain the logical sense)
![20191127_185643](https://user-images.githubusercontent.com/366258/69748016-c0031880-1147-11ea-9e3d-6bbe72b9d00e.jpg)

**Comment by ptrm** [comment]
> Ok, may be not generated hyphen from hyphenating words, but explicit hyphens, some example at https://unicode.org/reports/tr14/#Hyphen

Oh, true, that's a nice to have.

**Comment by poire-z** [comment]
Thanks for the screenshot. So, only `a i o u w  z` plain? and not `e` ? And not their diacritics variations like `ó ą ę` ?
> that's a nice to have

But how painful to have not? :)

**Comment by ptrm** [comment]
> Thanks for the screenshot. So, only `a i o u w z` plain? and not `e` ? And not their diacritics variations like `ó ą ę` ?

There are simply no such cases in real life ;) "i" is what "y" is in Spanish, I think. So implementing a catch-all would not reveal it's a catch-all, anyway ;)

> But how painful to have not? :)

I actually thought the hyphen thing doesn't occur a lot, but it turns out in my current book (fiction) it does, and I think in general too. I would say the sense will be obvious anyway, I would mind if the hyphen was lacking at the start of the line, but then I'm a nit-picking type in this matter. It's the auto-hyphenation and font display settings mostly that made me move from Kindles and Nickles to koreader ;)

Still, maybe there are cases in some specialist literature where it would matter in a sense-changing way, but that's a guess.

**Comment by ptrm** [comment]
> 320 should help with really rare cases

Right, edge cases, I checked the attached library, no visible changes ;)

**Comment by robert00s** [comment]
> Thanks for the screenshot. So, only a i o u w z plain? and not e ? And not their diacritics variations like ó ą ę ?

> >that's a nice to have

> But how painful to have not? :)

`a i o u w z` are single word that mean:
a - and (variant)
i - and (variant)
o - about
u - at
w - in
z - with

`e` is not a single Polish word. Similar to `ą, ó, z, ź, ź`.

I know that may be very difficult to do but it would be great to have options:

[ ] ignore orphans at the end of line
[ ] prevent orphans at the end of line 
[ ] publishers orphans at the end of line

First option will ignore `&nbsp;` second will force move single letter to next line and third will leave like that is now.

In my option
A) 9
B) 4
C) 5

**Comment by poire-z** [comment]
Thanks.

And there are no other single letter "word" or cases in polish, where this rule should _not_ / would _not_ need to be obeyed? Like if a book names its characters `Mister A and Miss Z meet in the street`, having the A or Z on next line and Mister or Miss and end of previous line would feel strange ? :)

> First option will ignore `&nbsp; `

You based these options on the current Orphans & Widows lines tweaks - but would there really be a use-case to ignore all `&nbsp;` ? Or a use-case to ignore only those after these single letters `a i o u w z` in polish text?

(But yes, adding and using custom css properties to let the user tweak that kind of things would be a good option to not bother with detecting the language and making hard rules - if we were not so near to the max number of css properties we can support (58/64).)

Another option would be to base these language tweaks on the hyphenation dict used. Do you both use the Polish hyphenation we ship? With polish text only, or even with english text? (I do use the French hyph with English text :) English hyph on english text feels bogus with my french habits :)

**Comment by ptrm** [comment]
> having the A or Z on next line and Mister or Miss and end of previous line would feel strange ? :)

That's too much for me :D It think it would not hurt to have it moved to another line.

> With polish text only, or even with english text?

I use auto picked hyphenation dicts, so respective to the document labguage. English broken according to Polish rules and vice versa is a mess ;)



This reminds me of another issue I see both on the publisher and engine side, with foreign words like "passe-partout" used in English or Polish: for me it should appear in html like `<span lang="fr">passe-partout</span>` and have French hyphenation applied ;) But since I have seen no such markup anywhere yet, I guess it's just a dream.

**Comment by poire-z** [comment]
I had some thoughts about handling that in https://github.com/koreader/crengine/issues/307#issuecomment-531739317. But indeed, if it's so rarely used, it would be much work for little benefit :|

**Comment by robert00s** [comment]
> And there are no other single letter "word" or cases in polish, where this rule should not / would not need to be obeyed? Like if a book names its characters Mister A and Miss Z meet in the street, having the A or Z on next line and Mister or Miss and end of previous line would feel strange ? :)

It's very difficult to create automatic all rules. `Mister A` or `Miss Z` are very rare. I agree that it would not hurt to have it moved to another line.

> You based these options on the current Orphans & Widows lines tweaks - but would there really be a use-case to ignore all &nbsp; ? Or a use-case to ignore only those after these single letters a i o u w z in polish text?

I prefere a use-case to ignore `&nbsp;` only after single letters. I have recently noticed that I prefer documents without `&nbsp;` when more letters are in a line and pages look better (that's just my opinion).

> Another option would be to base these language tweaks on the hyphenation dict used. Do you both use the Polish hyphenation we ship? With polish text only, or even with english text? (I do use the French hyph with English text :) English hyph on english text feels bogus with my french habits :)

I like that option :+1: 
I read almost only Polish books (fiction) with Polish hyphenation.

**Comment by poire-z** [comment]
I'm looking again at implementing some of the ideas from https://github.com/koreader/crengine/issues/307 , starting with using libunibreak and tweaking it depending on the hyphenation language selected (for now).

For the polish single word prepositions, I came up with some trick using hooks, replacing, for libunibreak, these single letters preceeded by a space with an opening paren `(`:
```c
lChar16 lb_char_sub_func_polish(lChar16 * text, int pos, int next_usable) {
    // https://github.com/koreader/koreader/issues/5645#issuecomment-559193057
    // Letters aiouwzAIOUWS are prepositions that should not be left at the
    // end of a line.
    // Make them behave (for libunibreak) just like a opening paren (which
    // being LBC_OP, will prevent a line break after it, even if followed
    // by a space).
    if ( pos >= 1 && text[pos-1] == ' ' ) {
        switch ( text[pos] ) {
            case 'A':
            case 'I':
            case 'O':
            case 'U':
            case 'W':
            case 'Z': // Meaning in english:
            case 'a': // and
            case 'i': // and
            case 'o': // about
            case 'u': // at
            case 'w': // in
            case 'z': // with
                return '(';
                break;
            default:
                break;
        }
    }
    return text[pos];
}
```
which seems to work:

<kbd>![image](https://user-images.githubusercontent.com/24273478/78123026-d844b600-740d-11ea-9308-e7bbe5670c60.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/78123077-edb9e000-740d-11ea-9309-ccfd16201a90.png)</kbd>

@ptrm @robert00s : does this idea/trick (`space + a|i|o|u|w|z` being handled like `space + (` in the matter of line breaking) _feels_ right ? Or can you think of cases where this would be wrong?

If anybody else reading this can think of cases in other languages where that king of trick may be needed, tell me (so I can see what kind of hook might be needed).

**Comment by poire-z** [comment]
Also, I'm trying to shorten https://github.com/adah1972/libunibreak/blob/master/src/linebreakdef.c and make it more generic.
We need a specific one for German (and also czech, slovak, sloven...), which get all quotes inverted vs the rest of the world :)

I think other languages would be fine with the one proposed for french and spanish:
```c
static const struct LineBreakProperties lb_prop_cre_Generic[] = {
    { 0x00AB, 0x00AB, LBP_OP }, /* Left double angle quotation mark: opening */   // SP FR RU
    { 0x00BB, 0x00BB, LBP_CL }, /* Right double angle quotation mark: closing */  // SP FR RU
    { 0x2018, 0x2018, LBP_OP }, /* Left single quotation mark: opening */         // EN SP FR CH
    { 0x201C, 0x201C, LBP_OP }, /* Left double quotation mark: opening */         // EN SP FR CH (NOT RU!!!)
    { 0x201D, 0x201D, LBP_CL }, /* Right double quotation mark: closing */        // EN SP FR CH
    { 0x2039, 0x2039, LBP_OP }, /* Left single angle quotation mark: opening */   // SP FR
    { 0x203A, 0x203A, LBP_CL }, /* Right single angle quotation mark: closing */  // SP FR
    { 0, 0, LBP_Undefined }
};

// German has most of these quote reverted
static const struct LineBreakProperties lb_prop_cre_German[] = {
    { 0x00AB, 0x00AB, LBP_CL }, /* Left double angle quotation mark: closing */
    { 0x00BB, 0x00BB, LBP_OP }, /* Right double angle quotation mark: opening */
    { 0x2018, 0x2018, LBP_CL }, /* Left single quotation mark: closing */
    { 0x2019, 0x2019, LBP_GL }, /* Right single quotation mark: glue */
    { 0x201C, 0x201C, LBP_CL }, /* Left double quotation mark: closing */
    { 0x2039, 0x2039, LBP_CL }, /* Left single angle quotation mark: closing */
    { 0x203A, 0x203A, LBP_OP }, /* Right single angle quotation mark: opening */
    { 0, 0, LBP_Undefined }
};

static const struct LineBreakProperties lb_prop_cre_Russian[] = {
    { 0x00AB, 0x00AB, LBP_OP }, /* Left double angle quotation mark: opening */
    { 0x00BB, 0x00BB, LBP_CL }, /* Right double angle quotation mark: closing */
    { 0x201C, 0x201C, LBP_CL }, /* Left double quotation mark: closing */
    { 0, 0, LBP_Undefined }
};
```
There's one proposed for Russian, which doesn't have some set for French & Spanish, which I think won't hurt if we add them. But there's this last line about U+201C https://www.fileformat.info/info/unicode/char/201c/index.htm - which says it might be used for "closing" in Russian (so, russians don't want it at start of line). @mergen3107 @hius07 @pkb @virxkane : is that true?

https://en.wikipedia.org/wiki/Quotation_mark
Oh my, finish acceps `»…»`, and polish is fine with both «…» and »…«... (@ptrm @robert00s is that right?) ... What a mess.

That's need for when the publisher has not set the appropriate `&nbsp;` - which happens often in french with `« some words »` - I have hardcoded some rules, which of course are wrong in german as noticed above.
Trying to do the right thing in all languages will probably be hard...

**Comment by poire-z** [comment]
Or may be all that is just needed when there are some (any kind of) space between the opening quote and the words - which seem to happen only in french (@pazos: what about spanish, which get the same rules than french?).
For other languages that don't use any space, the quotation mark will be with the quoted first word it is stuck to. Adding some rules might prevent it from being allowed to start a line and be first on the new line.

**Comment by poire-z** [comment]
And english/dutch don't expect to see `« … » `:
![image](https://user-images.githubusercontent.com/24273478/78128941-0bd80e00-7417-11ea-9d8f-5f54fa86d1a6.png)
Would they mind having the french rules for `before « inside » after ` ? Or it's safer to not include them, so in case there's some embedded inverted german quotes `before »inside« after`, no wrong decision is taken?

**Comment by hius07** [comment]
> There's one proposed for Russian, which doesn't have some set for French & Spanish, which I think won't hurt if we add them. But there's this last line about U+201C https://www.fileformat.info/info/unicode/char/201c/index.htm - which says it might be used for "closing" in Russian

No, it should not be used for closing. 201D is for closing.

**Comment by mergen3107** [comment]
@poire-z 
Yes, actually 201C is used as closing quotation mark in Russian, but only is the corresponding opening quotation mark is 201E. Let me summarize possibly combinations in Russian:
1) `«…»` aka `U+00AB` and `U+00BB`. This is common quotation mark, and it is always the first choice if there are nested quotation marks (I'll give and example after this).
2) `„…“` aka `U+201E` and `U+201C`. This is more frequent for classical literature.
3) `“…”` aka `U+201C` and `U+201D`. This can appear in the same fashion as option (1).
4) `‘…’` aka `U+2018` and `U+2019`. I've rarely seen these. 

By "nested quotation" I mean something like citing from a book that «Pushkin wrote his poem “Eugene Onegin” in 1825». Here the outer quotation pair is from option (1), the inner one is from option (2), but can also be from option (4). The inner one should never be (1), but I am not sure if publishing entities are strictly following these rules.

I have a suggestion on how to resolve all this mess about `»…»` and other weird combinations. What if we look at the sentence from its beginning, then find pairs of quotation marks, and whatever is the second mark in such pair must be treated as "closing"?

**Comment by hius07** [comment]
Oh, yes, (2) may be used for inner nested quotation, but it's rather obsolete.
The current standard is like here, §135-5:
http://www.evartist.narod.ru/text1/55.htm

**Comment by mergen3107** [comment]
Russian is an overly complicated language by itself, but it also tends somehow to be even more confusing when transitioning from written-on-paper to typography xD

**Comment by poire-z** [comment]
Thanks for the info.

> I have a suggestion on how to resolve all this mess about »…» and other weird combinations. What if we look at the sentence from its beginning, then find pairs of quotation marks, and whatever is the second mark in such pairs must be treated as "closing" ones, and thus should not appear on a new line by themselves?

Well, that would be quite a bit more costly. We currently (and also later if we go using libunibreak) just scan the string forward (no going backward).
I guess it's just ok to not include any specific rules for Russian (and actualy most languages), contrary to what I thought: in your samples, there is no space (never?) between the opening quote and first quoted letter (and between the last quoted letter and the closing quote), and that seems true even for nested quotes: so, there won't be any line break between these, but one will be allowed on the other side of the quote as there is a space there, and all will be fine.
(Unline French which has a space, and needs some rules.)

Another question about cyrillic: are there different glyphs/letter forms for the same cyrillic char depending on the language (russian vs ukrainian vs others?).
Was trying to see how my fonts would do if we give them the text language, but they all output the same glyphs, except one font (Vollkorn) with Bulgarian:
<kbd>![image](https://user-images.githubusercontent.com/24273478/78334574-47481900-758c-11ea-80c1-a63e89e188c1.png)</kbd>
(If you can think of any, even with other cyrillic based languages, please write a bit of input text that I can play with :)

**Comment by poire-z** [comment]
But we do currently have some (hardcoded) wrap prevention around some of these quotes, targetting my French.
You'd never witnessed/deplored that with your FB2 texts, large word spacing on a line because a quoted bit had to be pushed on next line?

**Comment by hius07** [comment]
Ukrainian letters are of same form as Russians absolutely.

**Comment by mergen3107** [comment]
`there is no space (never?) between the opening quote and first quoted letter`
Correct, there are no spaces on either sides.

`Another question about cyrillic: are there different glyphs/letter forms for the same cyrillic char depending on the language (russian vs ukrainian vs others?).`
In your examples you have some letter missing in Russian alphabet. Also I put here letters of Kazakh Ukrainian and Uzbek letters. Well, there is no such thing as `different glyphs/letter forms for the same cyrillic char`, but these derived letters are based on their Russian analogues, I placed them next to their origin by sound:
[pl]бвгджзйклмнпрстфхцчъьшщ аеёиоуыэюя[ru].[pl]
[pl]бвг**ғ**джзйк**қ**лмн**ң**прстфх**һ**цчъьшщ а**ә**еёио**ө**у**үұ**ы**і**эюя[kk].[pl]
[pl]бвг**ґ**джзйклмнпрстфхцчьшщ ае**є**и**ії**оуюя[uk].[pl]
[pl]бвг**ғ**джзйк**қ**лмнпрстфх**ҳ**цчъьшщ аеёиоу**ўү**ыэюя[uz].[pl]

I don't know what happened to Bulgarian letters in your example, it looks very strange, because Bulgarian should share at least the same Cyrillic letters, such as `в` (this is the second letter from the beginning of the line).

**Comment by mergen3107** [comment]
`You'd never witnessed/deplored that with your FB2 texts, large word spacing on a line because a quoted bit had to be pushed on next line?`
No, I don't recall such examples.

**Comment by poire-z** [comment]
> I don't know what happened to Bulgarian letters in your example, it looks very strange, because Bulgarian should share at least the same Cyrillic letters

https://www.lettersoup.de/what-shall-be-done-for-bulgarian-cyrillic-loclbgr/ says there are differences.
So, I guess it's really specific to Bulgarian, and I don't have to look for more anymore :)

**Comment by mergen3107** [comment]
Oh, wow! That's for the piece of enlightenment :)
It looks like as we go further, we discover more and more language-specific complexities and inconsistencies xD

**Comment by poire-z** [comment]
@Eduardomb22 : can you tell me if this is true, if you expect it (even with brazilian portuguese), if you miss it (and how much, on a pain scale of 0 to 10 :), and if you have found books that use that trick to get you duplicated hyphens:

https://unicode.org/reports/tr14/#Hyphen (that's about real hyphen/minus sign, not about hyphen added when hyphenating a word):

> The same convention is used in Portuguese, where the use of hyphens is common, because they are mandatory for verb forms that include a pronoun. Homographs or ambiguity may arise if hyphens are treated incorrectly: for example, "disparate" means "folly" while "dispara-te" means "fire yourself" (or "fires onto you"). Therefore the former needs to be line broken as
> 
>     dispara-
>     te
> 
> and the latter as
> 
>     dispara-
>     -te.
> 
> A recommended practice is to type <SHY, NON-BREAKING HYPHEN> instead of <HYPHEN> to achieve promotion of the hyphen to the next line. This practice is reportedly already common and supported by major text layout applications.

**Comment by mergen3107** [comment]
Back to `You'd never witnessed/deplored that with your FB2 texts, large word spacing on a line because a quoted bit had to be pushed on next line?` :
![Reader_2020-Apr-10_174618](https://user-images.githubusercontent.com/6279855/79027495-6fbaaf00-7b5a-11ea-83a4-a174f058eba0.png)
The two words in the end of the highlight are definitely force-pushed to the next line instead of being hyphenated.

**Comment by mergen3107** [comment]
This is on KOReader Stable v2020.03.2 on Kindle PW3 (5.9.7)

**Comment by Eduardomb22** [comment]
@poire-z Yes, it is true in brazilian portuguese.
I don´t miss it. I read mostly law books at my kindle and they don´t usually use this kind of verbal form.
I will try to find some books that have it.

**Comment by poire-z** [comment]
> The two words in the end of the highlight are definitely force-pushed to the next line instead of being hyphenated.

Can you check again if it really never get hyphenated if you change a bit the layout properties (font size, L/R margins) ? May be you're just near the point the quote+first part would not fit, so the whole thing is pushed to next line (where does "cpedHeMy" is allowed to be hyphenated?).

In my french test case, with that quote with or without a space, it gets hyphenated:
<kbd>![image](https://user-images.githubusercontent.com/24273478/79037115-9ca8a980-7bce-11ea-8648-95594339f29e.png)</kbd>

**Comment by mergen3107** [comment]
OK, sorry! False alarm :) I decreased the font by just 0.5 and did hyphenate at the first possible hyphenation. I also noticed another good example of quoted word which got hyphenated properly:
![Reader_2020-Apr-11_024503](https://user-images.githubusercontent.com/6279855/79037421-bc27de00-7b9e-11ea-8df8-ba1f0f4806d5.png)
